library/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/recorder/ScreenRecorderTest.kt

```diff
@@ -51,234 +59,317 @@ internal class ScreenRecorderTest {
+            verify(it.decorView.viewTreeObserver).addOnDrawListener(captor.capture())
+            verify(it.decorView.viewTreeObserver).removeOnDrawListener(captor.firstValue)
+        }
+        assertThat(testedRecorder.windowsListeners).isEmpty()
     }
 
     @Test
-    fun `M register the RecorderWindowCallback W startRecording{default callback is null}`(
-        forge: Forge
-    ) {
+    fun `M unregister and clean the listeners W stopRecording()`() {
         // Given
-        val mockWindow: Window = mock()
-        val mockActivity = mockActivity(
-            forge,
-            window = mockWindow,
-            defaultWindowCallback = null
-        )
+        testedRecorder.startRecording(fakeWindowsList, mockActivity)
 
         // When
-        testedRecorder.startRecording(mockActivity)
+        testedRecorder.stopRecording()
 
         // Then
-        val captor = argumentCaptor<Window.Callback>()
-        verify(mockWindow).callback = captor.capture()
-        assertThat(captor.firstValue).isInstanceOf(RecorderWindowCallback::class.java)
-        assertThat((captor.firstValue as RecorderWindowCallback).wrappedCallback)
-            .isInstanceOf(NoOpWindowCallback::class.java)
+        fakeWindowsList.forEach {
+            val captor = argumentCaptor<ViewTreeObserver.OnDrawListener>()
+            verify(it.decorView.viewTreeObserver).addOnDrawListener(captor.capture())
+            verify(it.decorView.viewTreeObserver).removeOnDrawListener(captor.firstValue)
+        }
+        assertThat(testedRecorder.windowsListeners).isEmpty()
     }
 
     @Test
-    fun `M do nothing W startRecording() { activity window is null }`() {
+    fun `M unregister first and clean the listeners W startRecording()`() {
         // Given
-        val mockActivity: Activity = mock()
+        testedRecorder.startRecording(fakeWindowsList, mockActivity)
 
         // When
-        testedRecorder.startRecording(mockActivity)
+        testedRecorder.startRecording(fakeWindowsList, mockActivity)
 
         // Then
-        assertThat(testedRecorder.drawListeners).isEmpty()
+        fakeWindowsList.forEach {
+            val captor = argumentCaptor<ViewTreeObserver.OnDrawListener>()
+            it.decorView.viewTreeObserver.inOrder {
+                verify().addOnDrawListener(captor.capture())
+                verify().removeOnDrawListener(captor.firstValue)
+                verify().addOnDrawListener(captor.capture())
+            }
+        }
     }
 
     @Test
-    fun `M unregister the RecorderOnDrawListener W stopRecording()`(forge: Forge) {
+    fun `M register listener startRecording() { more activities }`(forge: Forge) {
         // Given
-        val mockViewTreeObserver: ViewTreeObserver = mock()
-        val mockActivity = mockActivity(forge, mockViewTreeObserver)
-        testedRecorder.startRecording(mockActivity)
+        val fakeWindowsActivityPairs: List<Pair<List<Window>, Activity>> = forge.aList {
```

Comment:
kotlin cannot infer the type here? given that types of `aMockedWindowsList()` and `aMockedActivity()` are known.

Reply:
let me check